docs/pallas/gpu/pipelining.md

```diff
@@ -0,0 +1,325 @@
+  wg_axis: str,
+  memory_thread_idx: int | None = None,
+)
+```
+
+There are a few arguments specific to this pipeline emitter, which are:
+- `num_compute_wgs` specifies how many compute warpgroups to use. The pipeline emitter always uses a single memory warpgroup, so in `plgpu.kernel` you should specify `num_threads=num_compute_wgs+1`.
+- `memory_registers` controls how many registers to allocate to the memory warpgroup. The remaining registers are partitioned evenly among the compute warpgroups.
+- `wg_axis` the name of the warpgroup axis (as specified by the `thead_name` argument of `plgpu.kernel`).
+- `memory_thread_idx` specifies which warpgroup to designate as the memory warpgroup. Defaults to the last warpgroup.
+- `carry_coroutine` is a Python coroutine that enables you to carry values in registers through the pipeline. The coroutine runs only on the compute warpgroups and defines the initialization of the carry and consumption of the final carry. All compute warpgroup specific arrays should be instantiated here so the memory warpgroup does not materialize them in registers -- otherwise, you may experience slowdowns due to register spills.
+
+The kernel body of the warp specialized pipeline is run in parallel by all compute warpgroups, and SMEM is shared between compute warpgroups since they are scheduled within the same CUDA block.`lax.axis_index` can be used inside the kernel to obtain the warpgroup ID in order to divide up work amongst compute warpgroups.
+
+<!-- #endregion -->
+
+<!-- #region id="ZGbK5gIvFZKy" -->
+### Example: Matrix Multiplication with Warp Specialization
+
+The following example extends the previous matrix multiplication example to use warp specialization. This particular kernel uses 2 compute warpgroups, which operate on separate columns of the RHS matrix but share the same LHS. Each invocation of the pipeline therefore computes 2 adjacent blocks in the output matrix.
+
+<!-- #endregion -->
+
+<!-- #region id="NYWBqa9-bp2p" -->
+
+![matmul_ws](../../_static/pallas/gpu/pipeline_matmul_ws.svg)
+
+<!-- #endregion -->
+
+<!-- #region id="OkWmfqn7b53M" -->
+We use the `carry_coroutine` pattern to initialize the WGMMA accumulator, and copy the final accumulator from registers into SMEM. Here, the carry coroutine is defined in the function `compute_thread`. It is critical that the accumulator be created inside of the `compute_thread` function to avoid allocating it in the memory warpgroup which would waste registers. To perform the. WGMMA, we wrap the `wgmma` instruction in a `pl.run_state` in order to create an accumulator ref that is initialized to the carry value.
+
+Instead of using `pl.pallas_call` to call the kernel, we instead use the GPU-specific `plgpu.kernel` entry point. `plgpu.kernel` allows us to specify the number of warpgroups to launch per CUDA block via the `num_threads` argument, and allows us to specify a `thread_name` we can use to query the warpgroup index inside of the kernel.
```

Comment:
Please please please stop talking about warpgroups this much. It can be helpful to mention it from time to time, but we really should talk more The warpgroup really is not visible and not a concept that makes sense in the semantics of Pallas:MGPU

Reply:
I'll refer to this as "Mosaic thread" since there's potential for ambiguity between "CUDA thread" and "Mosaic thread" as Christos noted above.

Reply:
Done - renaming to "Pallas thread".